### PR TITLE
[PT] Extractor for fused convolution

### PR DESCRIPTION
**Pull request by AlexanderDokuchaev**
### Changes

- Get layer attributes by args, kwargs for `conv`, `batch_norm`, `group_norm` and `linear` functions.
- Extract submodules for FBC for custom `conv` and `batch norm` modules.
- Add `model_graph_manager` module to works with models with wrapped with `trace_parameters=True`. (model_analizer.py will be deprecated after finished refactor PTQ to use graph with constant nodes)
- Moved nncf/quantization/algorithms/weight_compression/torch_backend.py to model_graph_manager

### Related tickets

129581

### Tests

tests/torch/test_model_graph_manager.py
tests/torch/test_extractor.py

### CODE REVIEW

File: nncf/torch/extractor.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
+# Copyright (c) 2024 Intel Corporation
```
Comment (alexsu52) [inline nncf/torch/extractor.py:1]: Common comments to the extractor:
- It look like we need API to get a model by subgraph of NNCFGraph. Subgraph is defined by inputs nodes and outputs nodes. Thus, I suggest to have such API `extract_model(model, input_nodes, output_nodes) `
- What different between `ExtractedConv` and `ExtractedConvTranspose`? Could you make general module?

Let's discuss offline extractor API.
сс` @andrey-churkin

File: nncf/torch/dynamic_graph/wrappers.py [hunk 2]
```diff
@@ -186,7 +186,17 @@ def _execute_op(
...
+            from nncf.torch.nncf_network import NNCFNetwork
+
+            if (
+                isinstance(ctx._threading.thread_local.base_module_replica, NNCFNetwork)
+                and ctx._threading.thread_local.base_module_replica.nncf.trace_parameters
+            ):
+                layer_attrs = get_layer_attributes_from_args_and_kwargs(op_name, args, kwargs)
+                ignored_algos = []  # Ignored args is not defined for not replaced modules
+            else:
+                layer_attrs, ignored_algos = _collect_module_attrs_and_ignored_algorithms(ctx, op_name, args, kwargs)
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/torch/dynamic_graph/wrappers.py]: done

File: nncf/torch/extractor.py (hunk 1 above)
Comment (AlexanderDokuchaev) [reply nncf/torch/extractor.py:1]: 1. Changed signature to 
```
def extract_model(model: NNCFNetwork, input_nodes: List[str], output_nodes: List[str]) -> Optional[nn.Module]:
```

3. `convNd` and `conv_transposeNd` have different list of arguments.

File: nncf/torch/dynamic_graph/wrappers.py [hunk 3]
```diff
@@ -205,18 +205,25 @@ def _collect_module_attrs_and_ignored_algorithms(
...
-        curr_module = ctx.get_current_module()
-        if curr_module is None:
-            raise nncf.ValidationError(
-                f"Operation {op_name} requires module attributes, but it was executed outside any module"
-            )
-        layer_attrs = get_layer_attributes_from_module(curr_module, op_name)
-        if isinstance(curr_module, _NNCFModuleMixin):
-            ignored_algos = deepcopy(curr_module.ignored_algorithms)
-    elif op_name in OP_NAMES_REQUIRING_ATTRS_FROM_ARGS_KWARGS:
+    if (
```
Comment (alexsu52) [inline nncf/torch/dynamic_graph/wrappers.py:210]: Do you have a summary table of what metatypes are supported layer attributes collection if trace_parameters == True and False?

File: nncf/torch/extractor.py (hunk 1 above)
Comment (alexsu52) [reply nncf/torch/extractor.py:1]: > convNd and conv_transposeNd have different list of arguments.

You have the following forward:
```
        return getattr(F, self.fn_name)(
            input=x,
            weight=w,
            bias=self.bias,
            stride=self.layer_attributes.stride,
            padding=self.layer_attributes.padding_values,
            dilation=self.layer_attributes.dilations,
            groups=self.layer_attributes.groups,
        )
`` 
but you can use something like that:
```
        return getattr(F, self.fn_name)(*args, **kwarg))
```

File: nncf/torch/dynamic_graph/wrappers.py (hunk 3 above)
Comment (AlexanderDokuchaev) [reply nncf/torch/dynamic_graph/wrappers.py:210]: No, but it's align for both options.

File: nncf/torch/extractor.py (hunk 1 above)
Comment (AlexanderDokuchaev) [reply nncf/torch/extractor.py:1]: Changed to use one ExtractedFunc class

File: nncf/torch/dynamic_graph/wrappers.py (hunk 3 above)
Comment (alexsu52) [reply nncf/torch/dynamic_graph/wrappers.py:210]: Could you merge it If it's align?

File: nncf/torch/extractor.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        return getattr(F, self.fn_name)(input=x, **self.kwargs)
+
+
+def extract_conv(
+    input_node: NNCFNode,
+    output_node: NNCFNode,
+    model: NNCFNetwork,
+) -> ExtractedFunc:
+    """
+    Extracts a convolutional layer from an NNCF graph and constructs an ExtractedConv module.
```
Comment (alexsu52) [inline nncf/torch/extractor.py:69]: Please, add blank line.

Comment (AlexanderDokuchaev) [reply nncf/torch/extractor.py:69]: done

File: nncf/torch/dynamic_graph/wrappers.py (hunk 3 above)
Comment (AlexanderDokuchaev) [reply nncf/torch/dynamic_graph/wrappers.py]: done

File: nncf/torch/model_graph_manager.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+
+
+def _find_fq_node_in_constant_subgraph(node: NNCFNode, graph: NNCFGraph) -> Optional[NNCFNode]:
+    """
+    Finds a fake quantize node within a constant subgraph.
+
+    :param node: The starting node within the subgraph.
+    :param graph: The NNCFGraph containing the subgraph
+    :return: The found fake quantize node, or None if not found.
+    """
```
Comment (daniil-lyakhov) [inline nncf/torch/model_graph_manager.py:262]:
```suggestion
def _find_fq_node_in_constant_subgraph(start_node: NNCFNode, graph: NNCFGraph) -> Optional[NNCFNode]:
    """
    Finds a fake quantize node within a constant subgraph.

    :param start_node: The starting node within the subgraph.
    :param graph: The NNCFGraph containing the subgraph
    :return: The founded fake quantize node, or None if not found.
    """
```

File: nncf/torch/model_graph_manager.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+    if node.metatype == om.PTNoopMetatype:
+        prev_nodes = graph.get_previous_nodes(node)
+        if len(prev_nodes) != 1:
+            return None
+        return find_const_node_in_constant_subgraph(prev_nodes[0], graph)
```
Comment (daniil-lyakhov) [inline nncf/torch/model_graph_manager.py:267]: Should we look for fq nodes instead of const nodes?

File: nncf/torch/model_graph_manager.py [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+    if node.node_type in om.QUANTIZE_NODE_TYPES:
+        return node
+    return None
+
+
+def is_quantized_weights(node: NNCFNode, nncf_graph: NNCFGraph) -> bool:
+    """
+    Check that module have fake_quantizer for weight.
```
Comment (daniil-lyakhov) [inline nncf/torch/model_graph_manager.py:275]:
```suggestion
    Check that module have fake_quantizer for its weights.
```

File: nncf/torch/model_graph_manager.py [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+        id = PreHookId(op_address=op_addr, input_port_id=port_id)
+        for call_hook in model.nncf._compressed_context._pre_hooks.get(id, {}).values():
+            if isinstance(call_hook, ExternalQuantizerCallHook):
+                storage = getattr(model.nncf, call_hook._storage_name)
+                return storage[call_hook._storage_key]
+    else:
+        for call_hook in model.nncf._compressed_context._post_hooks.get(op_addr, {}).values():
+            if isinstance(call_hook, ExternalQuantizerCallHook):
+                storage = getattr(model.nncf, call_hook._storage_name)
+                return storage[call_hook._storage_key]
```
Comment (daniil-lyakhov) [inline nncf/torch/model_graph_manager.py:313]: Please refactor and decrease amount of duplicated code

**Review by daniil-lyakhov — CHANGES_REQUESTED** [review]
Looks good overall, please address my comments

File: nncf/torch/extractor.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+                    " module that contain batchnorm operator should be inhered from one of {BATCH_NORM_CLASSES}."
+                )
+                return None
+            return nn.Sequential(extracted_module, extracted_bn)
+    return extracted_module
+
+
+def extract_model(model: NNCFNetwork, input_nodes: List[str], output_nodes: List[str]) -> Optional[nn.Module]:
+    """
+    Extracts a submodule from a given NNCF network containing only the nodes from the input to the output node.
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/torch/extractor.py:193]: It's common function to extract sub model, from input to output nodes but currently it can extract only conv+bn

File: nncf/torch/model_graph_manager.py [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+
+    :param node: The NNCF node for which to find the constant input node.
+    :param port_id: The ID of the input port to consider.
+    :param graph: The NNCF graph containing the nodes.
+
+    :return: The NNCF node providing the constant input to the specified port, or None if no such node is found.
+    """
+    for prev_node in graph.get_previous_nodes(node):
+        edge = graph.get_edge(prev_node, node)
+        if edge.input_port_id == port_id:
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/torch/model_graph_manager.py:72]: yes, nncfgraph does not have api to get node by port_id

File: nncf/torch/model_graph_manager.py [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+        const.data = data
+    else:
+        setattr(module, const_attr_name, data)
+
+
+def set_const_data_to_port_id(data: torch.Tensor, node: NNCFNode, port_id: int, model: NNCFNetwork) -> None:
+    """
+    Sets the value of a constant tensor within a specified node in an NNCFNetwork.
+
+    :param data: The tensor containing the new value to be set for the constant.
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/torch/model_graph_manager.py:242]: Thinks that no, it's responsibility of code that will use this function.

File: nncf/torch/model_graph_manager.py [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+def is_quantized_weights(node: NNCFNode, nncf_graph: NNCFGraph) -> bool:
+    """
+    Check that module have fake_quantizer for weight.
+
+    :param node: The target node.
+    :param nncf_graph: The NNCF graph.
+    :return bool: return `True` if the node is quantized.
+    """
+    for edge in nncf_graph.get_input_edges(node):
+        if edge.input_port_id in node.metatype.weight_port_ids and _find_fq_node_in_constant_subgraph(
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/torch/model_graph_manager.py:288]: Returns True if at least one of the weights is quantized

File: nncf/torch/extractor.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/torch/extractor.py:193]: Can we call the function `extract_module` then?

**Review by alexsu52 — COMMENTED** [review]
Please, provide e2e results for PyTorch backend.

File: nncf/torch/model_graph_manager.py [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+            edge.from_node, nncf_graph
+        ):
+            return True
+    return False
```
Comment (daniil-lyakhov) [inline nncf/torch/model_graph_manager.py:292]:
```suggestion
    for edge in nncf_graph.get_input_edges(node):
        if edge.input_port_id in node.metatype.weight_port_ids and not _find_fq_node_in_constant_subgraph(
            edge.from_node, nncf_graph
        ):
            return False
    return True
```

**Review by daniil-lyakhov — APPROVED** [review]

**Review by alexsu52 — APPROVED** [review]
LGTM
